cmd/agent/app/reporter/client_metrics_test.go

```diff
@@ -0,0 +1,270 @@
+		assert.Len(t, tr.mr.ZipkinSpans(), 1)
+	})
+}
+
+func TestClientMetricsReporter_Jaeger(t *testing.T) {
+	testClientMetrics(func(tr *clientMetricsTest) {
+		blank := ""
+		clientUUID := "foobar"
+		nPtr := func(v int64) *int64 { return &v }
+		const prefix = "client_stats."
+		tag := func(name, value string) map[string]string { return map[string]string{name: value} }
+
+		tests := []struct {
+			clientUUID  *string
+			seqNo       *int64
+			stats       *jaeger.ClientStats
+			runExpire   bool // invoke expireClientMetrics to update the gauge
+			expLog      string
+			expCounters []metricstest.ExpectedMetric
+			expGauges   []metricstest.ExpectedMetric
+		}{
+			{},
+			{clientUUID: &blank},
+			{clientUUID: &clientUUID},
+			{
+				clientUUID: &clientUUID,
+				seqNo:      nPtr(100),
+				expLog:     clientUUID,
+				runExpire:  true,
+				expCounters: []metricstest.ExpectedMetric{
+					{Name: prefix + "batches_sent", Value: 100},
+				},
+				expGauges: []metricstest.ExpectedMetric{
+					{Name: prefix + "connected_clients", Value: 1},
+				},
+			},
+			{
+				clientUUID: &clientUUID,
+				seqNo:      nPtr(101),
+				expCounters: []metricstest.ExpectedMetric{
+					{Name: prefix + "batches_sent", Value: 101},
+				},
+			},
+			{clientUUID: &clientUUID, seqNo: nPtr(90)},
+			{
+				clientUUID: &clientUUID,
+				seqNo:      nPtr(110),
+				stats: &jaeger.ClientStats{
+					FullQueueDroppedSpans: 5,
+					TooLargeDroppedSpans:  6,
+					FailedToEmitSpans:     7,
+				}, expCounters: []metricstest.ExpectedMetric{
+					{Name: prefix + "batches_sent", Value: 110},
+					{Name: prefix + "spans_dropped", Tags: tag("cause", "full-queue"), Value: 5},
+					{Name: prefix + "spans_dropped", Tags: tag("cause", "too-large"), Value: 6},
+					{Name: prefix + "spans_dropped", Tags: tag("cause", "send-failure"), Value: 7},
+				},
+			},
+		}
+
+		for i, test := range tests {
+			t.Run(fmt.Sprintf("iter%d", i), func(t *testing.T) {
+				tr.logs.TakeAll()
+
+				batch := &jaeger.Batch{
+					Spans: []*jaeger.Span{{}},
+					Process: &jaeger.Process{
+						ServiceName: "blah",
+					},
+					SeqNo: test.seqNo,
+					Stats: test.stats,
+				}
+				if test.clientUUID != nil {
+					batch.Process.Tags = []*jaeger.Tag{{Key: "client-uuid", VStr: test.clientUUID}}
+				}
+
+				err := tr.r.EmitBatch(batch)
+				assert.NoError(t, err)
+				assert.Len(t, tr.mr.Spans(), i+1)
+
+				tr.assertLog(t, "new client", test.expLog)
+
+				if test.runExpire {
+					tr.r.expireClientMetrics(time.Now())
+				}
+
+				if m := test.expCounters; len(m) > 0 {
+					tr.mb.AssertCounterMetrics(t, m...)
+				}
+				if m := test.expGauges; len(m) > 0 {
+					tr.mb.AssertGaugeMetrics(t, m...)
+				}
+			})
+		}
+	})
+}
+
+func TestClientMetricsReporter_Delta(t *testing.T) {
+	m := int64(math.MaxInt64)
+	tests := []struct {
+		old, new, delta int64
+	}{
+		{old: 5, new: 10, delta: 5},
+		{old: m - 10, new: m - 5, delta: 5},
+		{old: m - 5, new: 10, delta: 15},
```

Comment:
Just out of curiosity, do you expect a client to ever reach `math.MaxInt64` without restarts?

Reply:
Perhaps you're right, we could ignore that possibility and simplify the code. Napkin math:
  * A service with 10k QPS => 1,000 batches / sec
  * this is about 300,000,000 years to roll over 2^63